### ISSUE TITLE
Use bindings for node.js

### ISSUE THREAD

**Issue by szmarczak** [yoshihitoh/zstd-codec#8]
emscripten is cool, but wouldn't it be much ~~better~~ faster to use bindings for node.js?

**Comment by yoshihitoh** [comment]
Hello @szmarczak ,
Sorry for replying so late.

> emscripten is cool, but wouldn't it be much better faster to use bindings for node.js?
I agree with that  too.
If running on Node.js and the performance is important, bindings are better choice.

IMO, Emscripten is suitable for browsers, hope that WebAssembly solve the performance issues.

**Comment by szmarczak** [comment]
> Sorry for replying so late.

No worries, that's ok 😃 

> IMO, Emscripten is suitable for browsers, hope that WebAssembly solve the performance issues.

Indeed. Nowadays all modern browsers run WebAssembly, for example Chrome or Firefox. We should use this.

So, what about the bindings? Is it in the plans?

**Comment by szmarczak** [comment]
@yoshihitoh Any news?

**Comment by yoshihitoh** [comment]
@szmarczak 

> So, what about the bindings? Is it in the plans?

`bindings` means binding for WebAssembly, is my understanding correct?
I'm planning to support WebAssembly in the next version, but that takes a little more time.

There are bindings for native libraries, so I will not support that.
- [node-zstandard](https://www.npmjs.com/package/node-zstandard)
- [node-zstd](https://www.npmjs.com/package/node-zstd)

**Comment by szmarczak** [comment]
@yoshihitoh You misunderstood me. I meant native bindings. Have you looked at when these repos were last published? They're not active anymore, outdated.

**Comment by yoshihitoh** [comment]
Okay, ic what bindings means.

I didn't realized that above bindings area outdated.
It's okay to support native bindings, but I don't know how to coexist native bindings and Emscriptened module.
Maybe the native binding will be another repository.

**Comment by szmarczak** [comment]
> Maybe the native binding will be another repository.

I think that would be the best choice. We can try something like this:

```
first repo:
node -> use bindings

second repo:
javascript -> check if browser supports wasm -> yes, use it
                           |
                          \|/
                    no, use emscripten
```

**Comment by yoshihitoh** [comment]
Okay.
I'm going to create new repo for native binding.
This is the first time to create native extension of Node.js, so need a little more time.

I'm not fluent in JavaScript.
If you know compression library that has sophisticated API, please let me know.

**Comment by szmarczak** [comment]
> This is the first time to create native extension of Node.js, so need a little more time.

Don't hurry 😃 That should be lower priority then, because this repo is the most important one.

> I'm not fluent in JavaScript.

I'm willing to help.

> If you know compression library that has sophisticated API, please let me know.

I think I don't understand you well. What's the goal? IMO it should support the most important features: compression, decompression, streams and dictionaries.

**Comment by yoshihitoh** [comment]
@szmarczak  Thanks! 😃 

> I think I don't understand you well. What's the goal? IMO it should support the most important features: compression, decompression, streams and dictionaries.

I want to know better API design in JavaScript.
Currently, zstd-codec uses normal function style like C/C++.
Pass the data to compress/decompress, and return the compressed/decompressed bytes.

[node-zstd](https://github.com/zwb-ict/node-zstd) and [node-zstandard](https://github.com/albertdb/node-zstandard) uses callback based style.

I don't know which style is better  😔

**Comment by szmarczak** [comment]
> I don't know which style is better 😔

There are some possibilities:
`compress(data, callback)` - this style is IMO the best solution, because Node.JS is designed to be asynchronous; also this type of function can be promisified.

`compress(data)` - IO blocking here is a big disadventage. Imagine compressing large amount of data... This would take a lot of time; maybe someone wants to do something in parallel while it's compressing.

I recommend you the first solution :)

**Comment by yoshihitoh** [comment]
Thanks for detailed solutions!
I'll try async style approach.

BTW I don't have enough time to development until June, updates will be following order.
1. zstd-codec 0.1.0 : enable WebAssembly    <-- May or June
2. zstd-codec 0.2.0 : support Node.js, (within this repo, to reduce code duplication)  <-- June or later

At the 0.2.0, split components into...
- core C++ module
- asm.js module (currently, using this module format)
- WebAssembly module
- native binding module (for Node.js)

`zstd-codec:0.2.0` will support ams.js/WebAssembly for browsers, and native binding for Node.js, if there is no technical issues.

**Comment by yoshihitoh** [comment]
Sorry for the behind schedule..

Progress:
- WebAssembly for Node.js : ok
- WebAssembly for Browsers: NOT YET

I thinks I can start next weekend...

**Comment by paambaati** [comment]
@yoshihitoh No hurries, but is there some progress here? Can I help with something?

**Comment by yoshihitoh** [comment]
@paambaati Sorry, I couldn't work on this issue 😢
WebAssembly version is already shipped, but Node.js bindings is not yet.

If you know well-architected JavaScript projects support both web and Node.js, please let me know.
I couldn't find out good example projects.

**Comment by yoshihitoh** [comment]
It is difficult to support both browser and Node.js when using Node native bidings because zstd-codec cannot determine the platform on build time.
So I'm planning to separate NPM packages into three
- zstd-codec (using asmjs, for browser.)
    - asm.js is compatible with JavaScript, so can run on many browsers
- zstd-codec-wasm (using WebAssembly, for browser)
- zstd-codec-node (using Native bindings, for Node.js)

To achieve above separation, I need to restructure project.
- zstd-codec-core: common module, shared C++ implementation only. (simple wrapper of zstd)
- zstd-codec-browser: browser specific components only. (bidings for Emscripten)
    - do nothing on build time (because zstd-codec includes asm.js/WebAssembly binaries.)
- zstd-codec-node: Node.js specific components only. (bindings for Node.js)
    - need to compile zstd and zstd-codec modules on build time

I have never build native extension for Node.js, but there is good example [node-zstd](https://github.com/zwb-ict/node-zstd) that will help me to implement the bindings.
I will apply these changes in `wip/restructure-projects` branch, but it will take time :)

**Comment by Erchoc** [comment]
> It is difficult to support both browser and Node.js when using Node native bidings because zstd-codec cannot determine the platform on build time. So I'm planning to separate NPM packages into three
> 
> * zstd-codec (using asmjs, for browser.)
>   
>   * asm.js is compatible with JavaScript, so can run on many browsers
> * zstd-codec-wasm (using WebAssembly, for browser)
> * zstd-codec-node (using Native bindings, for Node.js)
> 
> To achieve above separation, I need to restructure project.
> 
> * zstd-codec-core: common module, shared C++ implementation only. (simple wrapper of zstd)
> * zstd-codec-browser: browser specific components only. (bidings for Emscripten)
>   
>   * do nothing on build time (because zstd-codec includes asm.js/WebAssembly binaries.)
> * zstd-codec-node: Node.js specific components only. (bindings for Node.js)
>   
>   * need to compile zstd and zstd-codec modules on build time
> 
> I have never build native extension for Node.js, but there is good example [node-zstd](https://github.com/zwb-ict/node-zstd) that will help me to implement the bindings. I will apply these changes in `wip/restructure-projects` branch, but it will take time :)

does this feat done?
